execution/process.go

```diff
@@ -344,46 +357,74 @@ func (p *ProcessByKey) RunGarbageCollector(ctx context.Context, prefixedStorage
 		// We don't want to clear whole storage when sending MaxWatermark - on EndOfStream and WatermarkTrigger
 		if watermark != MaxWatermark {
 			// Collect every event time earlier than watermark - boundary (by default: watermark - 10min)
-			boundary := watermark.Add(time.Duration(-1*gbBoundary) * time.Second)
+			boundary := watermark.Add(-1 * gbBoundary * time.Second)
 
-			eventTimeMap := storage.NewMap(tx.WithPrefix(eventTimesSeenPrefix))
+			eventTimeSet := storage.NewMultiSet(tx.WithPrefix(eventTimesSeenPrefix))
 			octoEventTimeSlice := make([]octosql.Value, 0)
+			eventTimesBytesSlice := make([][]byte, 0)
+
+			var octoEventTime, keyTuple octosql.Value
 
-			var key, value octosql.Value
-			it := eventTimeMap.GetIterator()
-			for err := it.Next(&key, &value); err != storage.ErrEndOfIterator; err = it.Next(&key, &value) {
-				eventTime := key.AsTime()
+			// Iterating through event times seen
+			setIt := eventTimeSet.GetIterator()
+			for err := setIt.Next(&octoEventTime); err == nil; err = setIt.Next(&octoEventTime) {
+				eventTime := octoEventTime.AsTime()
 
 				if eventTime.Before(boundary) {
-					octoEventTimeSlice = append(octoEventTimeSlice, key)
+					eventTimeBytes := append(append([]byte("$"), octoEventTime.MonotonicMarshal()...), '$')
+
+					octoEventTimeSlice = append(octoEventTimeSlice, octoEventTime)
+					eventTimesBytesSlice = append(eventTimesBytesSlice, eventTimeBytes)
 
-					eventTimeBytes := append(append([]byte("$"), key.MonotonicMarshal()...), '$')
+					// Iterating through all keyTuples with that event time
+					keyTuplesSet := storage.NewMultiSet(tx.WithPrefix(eventTimesPrefixedSetsPrefix).WithPrefix(eventTimeBytes))
+
+					allKeysTuple := make([]octosql.Value, 0)
```

Comment:
MultiSet has a `ReadAll` function. Use it here please

Reply:
oooh yes, that's a new one, lovely